edb/schema/schema.py

```diff
@@ -1004,19 +1284,401 @@ def __init__(
+                to_type,
+                implicit=implicit,
+                assignment=assignment,
+            )
+        )
+
+    def get_casts_from_type(
+        self,
+        from_type: s_types.Type,
+        *,
+        implicit: bool = False,
+        assignment: bool = False,
+    ) -> FrozenSet[s_casts.Cast]:
+        return (
+            self._base_schema.get_casts_from_type(
+                from_type,
+                implicit=implicit,
+                assignment=assignment,
+            )
+            | self._top_schema.get_casts_from_type(
+                from_type,
+                implicit=implicit,
+                assignment=assignment,
+            )
+        )
+
+    def get_referrers(
+        self,
+        scls: so.Object,
+        *,
+        scls_type: Optional[Type[so.Object]] = None,
+        field_name: Optional[str] = None,
+    ) -> FrozenSet[so.Object]:
+        return (
+            self._base_schema.get_referrers(
+                scls,
+                scls_type=scls_type,
+                field_name=field_name,
+            )
+            | self._top_schema.get_referrers(
+                scls,
+                scls_type=scls_type,
+                field_name=field_name,
+            )
+        )
+
+    def get_referrers_ex(
+        self,
+        scls: so.Object,
+    ) -> Dict[
+        Tuple[Type[so.Object], str],
+        FrozenSet[so.Object],
+    ]:
+        base = self._base_schema.get_referrers_ex(scls)
+        top = self._top_schema.get_referrers_ex(scls)
+        return {
+            k: base.get(k, frozenset()) | top.get(k, frozenset())
+            for k in itertools.chain(base, top)
+        }
+
+    @overload
+    def get_by_id(
+        self,
+        obj_id: uuid.UUID,
+        default: Union[so.Object, so.NoDefaultT] = so.NoDefault,
+        *,
+        type: None = None,
+    ) -> so.Object:
+        ...
+
+    @overload
+    def get_by_id(  # NoQA: F811
+        self,
+        obj_id: uuid.UUID,
+        default: Union[so.Object_T, so.NoDefaultT] = so.NoDefault,
+        *,
+        type: Optional[Type[so.Object_T]] = None,
+    ) -> so.Object_T:
+        ...
+
+    @overload
+    def get_by_id(  # NoQA: F811
+        self,
+        obj_id: uuid.UUID,
+        default: None = None,
+        *,
+        type: None = None,
+    ) -> Optional[so.Object]:
+        ...
+
+    @overload
+    def get_by_id(  # NoQA: F811
+        self,
+        obj_id: uuid.UUID,
+        default: None = None,
+        *,
+        type: Type[so.Object_T],
+    ) -> Optional[so.Object_T]:
+        ...
+
+    def get_by_id(  # NoQA: F811
+        self,
+        obj_id: uuid.UUID,
+        default: Union[so.Object_T, so.NoDefaultT, None] = so.NoDefault,
+        *,
+        type: Optional[Type[so.Object_T]] = None,
+    ) -> Optional[so.Object_T]:
+        try:
+            return self._top_schema.get_by_id(
+                obj_id, type=type)
+        except errors.InvalidReferenceError:
```

Comment:
Maybe we optimize this one too by checking if the `_top_schema` has this id? Exceptions are expensive.